### ISSUE TITLE
Refactor processing pipeline to a monad form

### ISSUE THREAD

**Issue by andrus** [nhl/link-rest#215]
Using annotations for listeners is getting old. I am thinking of refactoring the LR processing pipeline using monad-like approach. This will require splitting `SelectContext` and friends into individual classes for every processing stage, and then using something simple like `Optional<T>` (or a custom monad) to build the pipeline:

```java
// default pipeline ... transformation goes like this:
// ClientRequest -> ParsedClientRequest -> ServerRequest -> CayenneRequest -> DataResponse

Optional<DataResponse<E2>> response =
   Optional.of(clientRequest)  // stage 0
      .flatMap(new ToParsedClientRequest<>()) // stage 1
      .flatMap(new ToServerRequest<>()) // stage 2
      .flatMap(new ToCayenneRequest<>()) // stage 3
      .flatMap(new ToDataResponse<>()); // stage 4
```
Building the pipeline on the fly (instead of caching a singleton) and separation of the state objects into individual classes gives us a great deal of flexibility.

If we want to add a custom stage to the default pipeline, we'd register a mapping function with `SelectBuilder`, that will be inserted in the right place in the pipeline behind the scenes. E.g.:

```java
// This will result in "flatMap(mapper)" call between stages 1 and 2
SelectBuilder.doAfterParse(Function<ParsedClientRequest<T>, ParsedClientRequest<T>> mapper);
```
Similarly if we want to add a custom stage that needs to execute itself and then skip the rest of the pipeline ("custom backend" use case), we'd register a function returning `Optional<DataResponse<T>>` : 

```java
// This will result in a "flatMap(mapper)" call after stage 1 .
// Stages 2 - 4 will be removed from the pipeline
SelectBuilder.finishAfterParse(Function<ParsedClientRequest<T>, DataResponse<T>> mapper);
```
No annotations are required, pipeline insertion position is identified by the method name. Mapper functions have specific in and out types and everything is very transparent.

**Comment by andrus** [comment]
This is a bigger task than what I initially thought, with many prerequisites. E.g. the approach above does not allow the functions to get access to the DI container. My latest ideas are committed to a [215_mondas branch](https://github.com/nhl/link-rest/tree/215_monads), and the endpoint code looks something like this:

```java
@GET
public SchemaDataAndEncoder<E2> getCustomPipeline(UriInfo uriInfo) {

    Schema<E2> schema = LR.idAndAttributesSchema(E2.class, config);
    SelectStageFactory<E2> stageFactory = LR.selectStages(config);

    return Stage.of(schema)
            .peek(s -> LOGGER.info("After pipeline started"))
            .next(stageFactory.queryParser(uriInfo.getQueryParameters()))
            .peek(s -> LOGGER.info("After query keys parsed"))
            .next(stageFactory.fetcher())
            .peek(s -> LOGGER.info("After data is loaded"))
            .next(stageFactory.encoder())
            .peek(s -> LOGGER.info("After encoder is installed"))
            .get();
}
```

It is actually compatible with the old-style request chains (the old chain can internally generate such code).

The important piece here is replacing `ResourceEntity` with an immutable `Schema` structure. So we start with a server-defined Schema (a combination of LrEntity and Constraint in our current terminology ...  also not unlike GraphQL) and generate derived schemas based on request protocol.

**Comment by andrus** [comment]
And here is the least invasive solution that is 100% backwards compatible and works with a mutable "context". This is based on the idea that there are currently two types of listeners: those that do not change the pipeline flow and those that do. The first category can be represented as `Consumer<SelectContext<T>>` (in case of select requests), and the second as `Function<SelectContext<T>, ProcessingStage<SelectContext<T>, T>>`. The first type can be passed via a `peek` method, and the second - as `flatMap`:

```java
LinkRest.select(Type.class)
	.uri(uri)
	.peek(Stage.AFTER_REQUEST_PARSED, context -> {}) // interceptor
	.flatMap(Stage.AFTER_REQUEST_PARSED, context -> new CustomStage())// router
	.get();
```

**Comment by andrus** [comment]
A less ambitious version of this task is implemented by #240 and friends.
